simulator/infra/cache/cache_tag_array.cpp

```diff
@@ -1,177 +1,105 @@
-                  << "less than size of block (size in bytes of cache should "
-                  << "be not less than number of ways and size of block in "
-                  << "bytes)." << critical;
-
-    /*
-     * It also checks "size_in_bytes % line_size != 0" and
-     * "size_in_bytes % ways != 0".
-     */
-    if ( ( size_in_bytes % ( line_size * ways)) != 0)
-        serr << "ERROR: Wrong arguments! Size of cache should be a "
-                  << "multiple of block size in bytes and number of ways."
-                  << critical;
-
-    /* The next two use: "2^a=b"<=>"b=100...000[2]"<=>"(b&(b-1))=0". */
-    if ( !is_power_of_two( size_in_bytes / ( line_size * ways)))
-        serr << "ERROR: Wrong arguments! Number of sets should be a power"
-                  << " of 2." << critical;
-
-    if ( !is_power_of_two( line_size))
-        serr << "ERROR: Wrong arguments! Block size should be a power of "
-                  << "2." << critical;
-}
 
-CacheTagArray::CacheTagArray( uint32 size_in_bytes,
-                              uint32 ways,
-                              uint32 line_size,
-                              uint32 addr_size_in_bits) :
-                              CacheTagArrayCheck( size_in_bytes, ways,
-                                                  line_size, addr_size_in_bits),
-                              num_sets( size_in_bytes / ( line_size * ways)),
-                              addr_mask( bitmask<Addr>( addr_size_in_bits)),
-                              array( num_sets, std::vector<CacheTag>( ways)),
-                              lru( ways, num_sets)
-{
+                 
 
+    if ( size_in_bytes < ways * size_of_line)
+        serr << "ERROR: Wrong arguments! Cache size should be greater"
+             << "than the number of ways multiplied by line size" 
+             << std::endl << critical;
+
+    if ( !is_power_of_two( size_in_bytes))
+        serr << "ERROR: Wrong argumets! Cache size should be a power of 2"
+             << std::endl << critical;
+    
+    if ( !is_power_of_two( size_of_line))
+        serr << "ERROR: Wrong arguments! Block size should be a power of 2"
+             << std::endl << critical;
+    
+    if ( size_in_bytes % ( size_of_line * ways) != 0)
+        serr << "ERROR: Wrong arguments! Cache size should be multiple of"
+             << "the number of ways and line size"
+             << std::endl << critical;
 }
 
+
+
 std::pair<bool, uint32> CacheTagArray::read( Addr addr)
 {
-    const auto lookup_result = read_no_touch( addr);
-    const auto&[ is_hit, way_num] = lookup_result;
+    uint32 num_set = set( addr);
+    uint32 num_tag = tag( addr);
+
+    auto result = read_no_touch( addr);
 
-    if ( is_hit)
-    {
-        const auto set_num = set( addr);
-        lru.touch( set_num, way_num); // update LRU info
-    }
+    if ( result.first)
+        lru_module.update( num_set, num_tag); // update LRU if it's a hit
```

Comment:
# of way should be used here, not a tag, right?

Reply:
No, we need a tag here.

Reply:
No, we need not. LRU class should know nothing about tags. It has only two basic operations:
* Put some way to the head of the list (== touch)
* Take a way from the tail ( == update == allocate)

Reply:
Then there should be a kind of reflection of WAYS to TAGS( e,g, another map which is not good), because we need somehow to know the least recently used tag to erase it from the hash table.

Reply:
Well, I will use another map.

Reply:
> Then there should be a kind of reflection of WAYS to TAGS( e,g, another map which is not good), because we need somehow to know the least recently used tag to erase it from the hash table.

Yes, it should be used for CacheTagArray. It is a common operation for cache ("check what data is in way 1 of set 4").

Reply:
> Well, I will use another map.

Note that for PseudoLRU there is no map inside LRU class, it has a different mechanism.

Reply:
I have used a map inside CacheTagArray.